File: golang/cmd/http/handlers.go
package main

import (
	"net/http"
	"strconv"

	"github.com/gin-gonic/gin"
)

func getItems(c *gin.Context) {
	c.IndentedJSON(http.StatusOK, items)
}

func createItem(c *gin.Context) {
	var newItem Item

	if err := c.BindJSON(&newItem); err != nil {
		return
	}

	newItem.ID = nextID
	nextID++
	items = append(items, newItem)

	c.IndentedJSON(http.StatusCreated, newItem)
}

func getItem(c *gin.Context) {
	id, err := strconv.Atoi(c.Param("id"))
	if err != nil {
		c.IndentedJSON(http.StatusBadRequest, gin.H{"message": "invalid item ID"})
		return
	}

	for _, a := range items {
		if a.ID == id {
			c.IndentedJSON(http.StatusOK, a)
			return
		}
	}
	c.IndentedJSON(http.StatusNotFound, gin.H{"message": "item not found"})
}

func updateItem(c *gin.Context) {
	id, err := strconv.Atoi(c.Param("id"))
	if err != nil {
		c.IndentedJSON(http.StatusBadRequest, gin.H{"message": "invalid item ID"})
		return
	}

	var updatedItem Item
	if err := c.BindJSON(&updatedItem); err != nil {
		return
	}

	for i, a := range items {
		if a.ID == id {
			updatedItem.ID = a.ID
			items[i] = updatedItem
			c.IndentedJSON(http.StatusOK, updatedItem)
			return
		}
	}
	c.IndentedJSON(http.StatusNotFound, gin.H{"message": "item not found"})
}

func deleteItem(c *gin.Context) {
	id, err := strconv.Atoi(c.Param("id"))
	if err != nil {
		c.IndentedJSON(http.StatusBadRequest, gin.H{"message": "invalid item ID"})
		return
	}

	for i, a := range items {
		if a.ID == id {
			items = append(items[:i], items[i+1:]...)
			c.IndentedJSON(http.StatusOK, gin.H{"message": "item deleted"})
			return
		}
	}
	c.IndentedJSON(http.StatusNotFound, gin.H{"message": "item not found"})
}

File: javascript/src/handlers.js
let items = [];
let nextId = 1;

const getItems = (req, res) => {
    res.json(items);
};

const createItem = (req, res) => {
    const item = { id: nextId++, ...req.body };
    items.push(item);
    res.status(201).json(item);
};

const getItem = (req, res) => {
    const item = items.find(i => i.id === parseInt(req.params.id));
    if (item) {
        res.json(item);
    } else {
        res.status(404).send('Item not found');
    }
};

const updateItem = (req, res) => {
    const index = items.findIndex(i => i.id === parseInt(req.params.id));
    if (index !== -1) {
        // Update everything except the id
        const updatedItem = { ...items[index], ...req.body, id: items[index].id };
        items[index] = updatedItem;
        res.json(updatedItem);
    } else {
        res.status(404).send('Item not found');
    }
};

const deleteItem = (req, res) => {
    const index = items.findIndex(i => i.id === parseInt(req.params.id));
    if (index !== -1) {
        items = items.filter(i => i.id !== parseInt(req.params.id));
        res.status(200).send('Item deleted');
    } else {
        res.status(404).send('Item not found');
    }
};

module.exports = { getItems, createItem, getItem, updateItem, deleteItem };